north/src/runtime/state.rs

```diff
@@ -407,74 +397,86 @@ impl State {
-                    warn!("Cannot uninstall started container {}", app);
-                    return Err(Error::ApplicationRunning(app.manifest().name.clone()));
-                }
-                Some(app)
-            }
-            None => self.resources.get(&(name.to_string(), version.clone())),
-        };
-        if let Some(app) = to_uninstall {
+        let (manifest, path) = installed_npk.unwrap();
+
+        let instances = manifest
+            .instances
+            .filter(|i| *i > 1)
+            .map(|i| {
+                (0..i)
+                    .into_iter()
+                    .map(|i| format!("{}-{:03}", manifest.name, i))
+                    .collect()
+            })
+            .unwrap_or_else(|| vec![manifest.name.to_string()]);
+
+        for app in instances.iter().filter_map(|name| {
+            self.applications
+                .get(name)
+                .or_else(|| self.resources.get(&(name.to_string(), version.clone())))
+        }) {
             if app.is_running() {
                 warn!("Cannot uninstall started container {}", app);
                 return Err(Error::ApplicationRunning(app.manifest().name.clone()));
             }
-            info!("Uninstalling {}", app);
-            umount_npk(app.container()).await.map_err(Error::Mount)?;
-            self.applications.remove(name);
-            self.resources.remove(&(name.to_string(), version.clone()));
 
-            // Remove npk from registry
-            for d in &self.config.directories.container_dirs {
-                let mut dir = fs::read_dir(&d)
-                    .await
-                    .map_err(|e| Error::Io(format!("Failed to read {}", d.display()), e))?;
-                while let Some(res) = dir.next().await {
-                    let entry = res.map_err(|e| {
-                        Error::Io(
-                            "Could not read directory".to_string(), // TODO: Which directory?
-                            e,
-                        )
-                    })?;
-                    let manifest = read_manifest(entry.path().as_path(), &self.signing_keys)
-                        .map_err(Error::Npk)?;
-
-                    if manifest.name == name && manifest.version == *version {
-                        fs::remove_file(&entry.path()).await.map_err(|e| {
-                            Error::Io(format!("Failed to remove {}", entry.path().display()), e)
-                        })?;
-                    }
-                }
-            }
+            info!("Unmounting {}", app);
+            umount_npk(app.container()).await.map_err(Error::Mount)?;
+        }
 
-            Self::notification(
-                &self.events_tx,
-                Notification::Uninstalled(name.to_owned(), version.clone()),
-            )
-            .await;
+        for instance in instances {
+            let key = (instance, version.clone());
+            self.applications.remove(&key.0);
+            self.resources.remove(&key);
```

Comment:
Not necessary. If there were any running instances, we would already have bailed.

Reply:
What is not necessary? we are uninstalling the container, we have to remove it from the state correct?

Reply:
We only have running instances if the app is running...and in this case we won't uninstall anyway

Reply:
I'm not sure if I'm understanding correctly. If there is any running instance, this par of the code wont get executed. Maybe the name "instances" is confusing.

Reply:
you are right... running != installed => so we need to remove the mounted apps that are not running